File: sdk/verifiers/rust/tests/canonical.rs
#![recursion_limit = "256"]

use pipelock_verifier_rs::canonical::{canonicalize_action_record, canonicalize_receipt};
use pipelock_verifier_rs::signing::verify_receipt_with_options;
use serde_json::json;
use sha2::{Digest, Sha256};

fn full_receipt() -> serde_json::Value {
    json!({
        "version": 1,
        "action_record": {
            "version": 1,
            "action_id": "ts-full-0001",
            "action_type": "write",
            "timestamp": "2026-05-10T12:34:56.789Z",
            "principal": "org:test",
            "actor": "agent:test",
            "delegation_chain": ["root", "delegate"],
            "target": "https://example.com/<x>&y",
            "intent": "post data",
            "data_classes_in": ["secret", "prompt"],
            "data_classes_out": ["summary"],
            "side_effect_class": "external_write",
            "reversibility": "compensatable",
            "policy_hash": "sha256:abc",
            "verdict": "warn",
            "session_taint_level": "medium",
            "session_contaminated": true,
            "recent_taint_sources": [{
                "url": "https://source.example/a",
                "kind": "prompt",
                "level": 5,
                "timestamp": "2026-05-10T12:34:55Z",
                "receipt_id": "r1",
                "match_reason": "pattern"
            }],
            "session_task_id": "task-1",
            "session_task_label": "review",
            "authority_kind": "operator",
            "taint_decision": "ask",
            "taint_decision_reason": "tainted input",
            "task_override_applied": true,
            "contract_winning_source": "manifest",
            "contract_live_verdict": "warn",
            "contract_policy_sources": ["policy-a", "policy-b"],
            "contract_rule_id": "rule-1",
            "active_manifest_hash": "sha256:manifest",
            "contract_hash": "sha256:contract",
            "contract_selector_id": "selector-1",
            "contract_generation": 7,
            "transport": "https",
            "method": "POST",
            "layer": "dlp",
            "pattern": "token",
            "severity": "warning",
            "redaction": {
                "profile": "default",
                "provider": "openai",
                "parser": "json",
                "total_redactions": 2,
                "by_class": { "token": 1, "secret": 1 },
                "cache_boundary_kept": true
            },
            "request_id": "req-1",
            "chain_prev_hash": "genesis",
            "chain_seq": 0,
            "venue": "test-venue",
            "jurisdiction": "test-jurisdiction",
            "rulebook_id": "rulebook-v1",
            "remedy_class": "notify",
            "contestation_window": "24h",
            "precedent_refs": ["p1", "p2"]
        },
        "signature": "ed25519:dc7bdb6220e7dd261ca6a55f295ee0ca44c8dbb04c36a07940ee11730c2119dd1bae6e96ea6d465a7c6ba357119c2218a795b2eec17f424d6e070e03b9c9540c",
        "signer_key": "7de2d117b21faaa0f1d9d3d02fcba13838bef0c75caddf71de376f0bb837bfbc"
    })
}

fn sha256(data: &[u8]) -> String {
    hex::encode(Sha256::digest(data))
}

#[test]
fn canonical_action_record_matches_go_hash_for_all_current_fields() {
    let receipt = full_receipt();
    assert_eq!(
        sha256(&canonicalize_action_record(&receipt["action_record"])),
        "8d5805f40a979a44983971f1a1a5de677cfa173edc33d71146c586a12a1ff3e1"
    );
}

#[test]
fn canonical_receipt_envelope_matches_go_hash() {
    let receipt = full_receipt();
    assert_eq!(
        sha256(&canonicalize_receipt(&receipt)),
        "1b07dab8572e98c5f823cfdc449cbce6711d6ed626df500d739fd9ba9b630345"
    );
}

#[test]
fn full_field_receipt_signature_verifies() {
    let receipt = full_receipt();
    verify_receipt_with_options(&receipt, "", true).expect("signature verifies");
}

// A segment-genesis receipt after a signing-key rotation carries a
// key_transition marker that the Go signer includes in the canonical preimage
// (after chain_seq, before venue). The verifier MUST reproduce it byte-for-byte
// or a rotated receipt fails to verify. Pinned to the Go-produced hash.
#[test]
fn canonical_action_record_with_key_transition_matches_go_hash() {
    let record = json!({
        "version": 1,
        "action_id": "ts-kt-0001",
        "action_type": "read",
        "timestamp": "2026-05-10T12:34:56.789Z",
        "principal": "org:test",
        "actor": "agent:test",
        "target": "https://example.com/x",
        "side_effect_class": "external_read",
        "reversibility": "full",
        "policy_hash": "sha256:abc",
        "verdict": "allow",
        "transport": "fetch",
        "chain_prev_hash": "priortail",
        "chain_seq": 0,
        "key_transition": {
            "prior_signer_key": "7de2d117b21faaa0f1d9d3d02fcba13838bef0c75caddf71de376f0bb837bfbc",
            "prior_chain_seq": 41,
            "prior_chain_hash": "priortail"
        }
    });
    assert_eq!(
        sha256(&canonicalize_action_record(&record)),
        "e50a5512f6571afdd0196315580707451ec81e9637e9fb51d988bb6c175b1b40"
    );
}

#[test]
fn canonical_action_record_preserves_explicit_empty_key_transition() {
    let record = json!({
        "version": 1,
        "action_id": "ts-kt-empty",
        "action_type": "read",
        "timestamp": "2026-05-10T12:34:56.789Z",
        "principal": "org:test",
        "actor": "agent:test",
        "target": "https://example.com/x",
        "side_effect_class": "external_read",
        "reversibility": "full",
        "policy_hash": "sha256:abc",
        "verdict": "allow",
        "transport": "fetch",
        "chain_prev_hash": "priortail",
        "chain_seq": 0,
        "key_transition": {}
    });
    let canonical =
        String::from_utf8(canonicalize_action_record(&record)).expect("canonical UTF-8");
    assert!(
        canonical.contains(
            r#""key_transition":{"prior_signer_key":"","prior_chain_seq":0,"prior_chain_hash":""}"#
        ),
        "canonical bytes should preserve and normalize explicit empty key_transition: {canonical}"
    );
}

File: internal/receipt/canonical.go
// Copyright 2026 Josh Waldrep
// SPDX-License-Identifier: Apache-2.0

package receipt

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/luckyPipewrench/pipelock/internal/session"
)

func canonicalActionRecord(version int, ar ActionRecord) ([]byte, error) {
	switch version {
	case ReceiptVersion:
		return canonicalActionRecordV1(ar)
	default:
		return nil, fmt.Errorf("unsupported receipt canonical version %d", version)
	}
}

// canonicalActionRecordV1 is the frozen signing projection for receipt version
// 1. Do not add future ActionRecord fields here. If the signed surface changes,
// create a new projection and select it from canonicalActionRecord.
func canonicalActionRecordV1(ar ActionRecord) ([]byte, error) {
	return json.Marshal(actionRecordCanonicalV1{
		Version:               ar.Version,
		ActionID:              ar.ActionID,
		ParentActionID:        ar.ParentActionID,
		ActionType:            ar.ActionType,
		Timestamp:             ar.Timestamp,
		Principal:             ar.Principal,
		Actor:                 ar.Actor,
		DelegationChain:       ar.DelegationChain,
		Target:                ar.Target,
		Intent:                ar.Intent,
		DataClassesIn:         ar.DataClassesIn,
		DataClassesOut:        ar.DataClassesOut,
		SideEffectClass:       ar.SideEffectClass,
		Reversibility:         ar.Reversibility,
		PolicyHash:            ar.PolicyHash,
		Verdict:               ar.Verdict,
		DecisionPhase:         ar.DecisionPhase,
		DeferID:               ar.DeferID,
		ResolutionPolicy:      ar.ResolutionPolicy,
		ResolutionSource:      ar.ResolutionSource,
		SessionID:             ar.SessionID,
		SessionIDOriginal:     ar.SessionIDOriginal,
		SessionTaintLevel:     ar.SessionTaintLevel,
		SessionContaminated:   ar.SessionContaminated,
		RecentTaintSources:    canonicalTaintSourceRefsV1(ar.RecentTaintSources),
		SessionTaskID:         ar.SessionTaskID,
		SessionTaskLabel:      ar.SessionTaskLabel,
		AuthorityKind:         ar.AuthorityKind,
		TaintDecision:         ar.TaintDecision,
		TaintDecisionReason:   ar.TaintDecisionReason,
		TaskOverrideApplied:   ar.TaskOverrideApplied,
		ContractWinningSource: ar.ContractWinningSource,
		ContractLiveVerdict:   ar.ContractLiveVerdict,
		ContractPolicySources: ar.ContractPolicySources,
		ContractRuleID:        ar.ContractRuleID,
		ActiveManifestHash:    ar.ActiveManifestHash,
		ContractHash:          ar.ContractHash,
		ContractSelectorID:    ar.ContractSelectorID,
		ContractGeneration:    ar.ContractGeneration,
		Transport:             ar.Transport,
		Method:                ar.Method,
		Layer:                 ar.Layer,
		Pattern:               ar.Pattern,
		Severity:              ar.Severity,
		Redaction:             canonicalRedactionSummaryV1(ar.Redaction),
		Shield:                canonicalShieldSummaryV1(ar.Shield),
		RequestID:             ar.RequestID,
		ChainPrevHash:         ar.ChainPrevHash,
		ChainSeq:              ar.ChainSeq,
		RunNonce:              ar.RunNonce,
		KeyTransition:         canonicalKeyTransitionV1(ar.KeyTransition),
		Venue:                 ar.Venue,
		Jurisdiction:          ar.Jurisdiction,
		RulebookID:            ar.RulebookID,
		RemedyClass:           ar.RemedyClass,
		ContestationWindow:    ar.ContestationWindow,
		PrecedentRefs:         ar.PrecedentRefs,
	})
}

type actionRecordCanonicalV1 struct {
	Version int `json:"version"`

	ActionID       string     `json:"action_id"`
	ParentActionID string     `json:"parent_action_id,omitempty"`
	ActionType     ActionType `json:"action_type"`
	Timestamp      time.Time  `json:"timestamp"`

	Principal       string   `json:"principal"`
	Actor           string   `json:"actor"`
	DelegationChain []string `json:"delegation_chain"`

	Target string `json:"target"`

	Intent         string   `json:"intent,omitempty"`
	DataClassesIn  []string `json:"data_classes_in,omitempty"`
	DataClassesOut []string `json:"data_classes_out,omitempty"`

	SideEffectClass SideEffectClass `json:"side_effect_class"`
	Reversibility   Reversibility   `json:"reversibility"`

	PolicyHash string `json:"policy_hash"`
	Verdict    string `json:"verdict"`

	DecisionPhase       string                      `json:"decision_phase,omitempty"`
	DeferID             string                      `json:"defer_id,omitempty"`
	ResolutionPolicy    string                      `json:"resolution_policy,omitempty"`
	ResolutionSource    string                      `json:"resolution_source,omitempty"`
	SessionID           string                      `json:"session_id,omitempty"`
	SessionIDOriginal   string                      `json:"session_id_original,omitempty"`
	SessionTaintLevel   string                      `json:"session_taint_level,omitempty"`
	SessionContaminated bool                        `json:"session_contaminated,omitempty"`
	RecentTaintSources  []taintSourceRefCanonicalV1 `json:"recent_taint_sources,omitempty"`
	SessionTaskID       string                      `json:"session_task_id,omitempty"`
	SessionTaskLabel    string                      `json:"session_task_label,omitempty"`
	AuthorityKind       string                      `json:"authority_kind,omitempty"`
	TaintDecision       string                      `json:"taint_decision,omitempty"`
	TaintDecisionReason string                      `json:"taint_decision_reason,omitempty"`
	TaskOverrideApplied bool                        `json:"task_override_applied,omitempty"`

	ContractWinningSource string   `json:"contract_winning_source,omitempty"`
	ContractLiveVerdict   string   `json:"contract_live_verdict,omitempty"`
	ContractPolicySources []string `json:"contract_policy_sources,omitempty"`
	ContractRuleID        string   `json:"contract_rule_id,omitempty"`
	ActiveManifestHash    string   `json:"active_manifest_hash,omitempty"`
	ContractHash          string   `json:"contract_hash,omitempty"`
	ContractSelectorID    string   `json:"contract_selector_id,omitempty"`
	ContractGeneration    uint64   `json:"contract_generation,omitempty"`

	Transport string                       `json:"transport"`
	Method    string                       `json:"method,omitempty"`
	Layer     string                       `json:"layer,omitempty"`
	Pattern   string                       `json:"pattern,omitempty"`
	Severity  string                       `json:"severity,omitempty"`
	Redaction *redactionSummaryCanonicalV1 `json:"redaction,omitempty"`
	Shield    *shieldSummaryCanonicalV1    `json:"shield,omitempty"`
	RequestID string                       `json:"request_id,omitempty"`

	ChainPrevHash string `json:"chain_prev_hash"`
	ChainSeq      uint64 `json:"chain_seq"`

	RunNonce      string                    `json:"run_nonce,omitempty"`
	KeyTransition *keyTransitionCanonicalV1 `json:"key_transition,omitempty"`

	Venue              string   `json:"venue,omitempty"`
	Jurisdiction       string   `json:"jurisdiction,omitempty"`
	RulebookID         string   `json:"rulebook_id,omitempty"`
	RemedyClass        string   `json:"remedy_class,omitempty"`
	ContestationWindow string   `json:"contestation_window,omitempty"`
	PrecedentRefs      []string `json:"precedent_refs,omitempty"`
}

type taintSourceRefCanonicalV1 struct {
	URL         string             `json:"url"`
	Kind        string             `json:"kind"`
	Level       session.TaintLevel `json:"level"`
	Timestamp   time.Time          `json:"timestamp"`
	ReceiptID   string             `json:"receipt_id,omitempty"`
	MatchReason string             `json:"match_reason,omitempty"`
}

func canonicalTaintSourceRefsV1(in []session.TaintSourceRef) []taintSourceRefCanonicalV1 {
	if len(in) == 0 {
		return nil
	}
	out := make([]taintSourceRefCanonicalV1, 0, len(in))
	for _, source := range in {
		out = append(out, taintSourceRefCanonicalV1{
			URL:         source.URL,
			Kind:        source.Kind,
			Level:       source.Level,
			Timestamp:   source.Timestamp,
			ReceiptID:   source.ReceiptID,
			MatchReason: source.MatchReason,
		})
	}
	return out
}

type keyTransitionCanonicalV1 struct {
	PriorSignerKey string `json:"prior_signer_key"`
	PriorChainSeq  uint64 `json:"prior_chain_seq"`
	PriorChainHash string `json:"prior_chain_hash"`
}

func canonicalKeyTransitionV1(in *KeyTransition) *keyTransitionCanonicalV1 {
	if in == nil {
		return nil
	}
	return &keyTransitionCanonicalV1{
		PriorSignerKey: in.PriorSignerKey,
		PriorChainSeq:  in.PriorChainSeq,
		PriorChainHash: in.PriorChainHash,
	}
}

type redactionSummaryCanonicalV1 struct {
	Profile           string         `json:"profile,omitempty"`
	Provider          string         `json:"provider,omitempty"`
	Parser            string         `json:"parser,omitempty"`
	TotalRedactions   int            `json:"total_redactions,omitempty"`
	ByClass           map[string]int `json:"by_class,omitempty"`
	CacheBoundaryKept bool           `json:"cache_boundary_kept,omitempty"`
}

func canonicalRedactionSummaryV1(in *RedactionSummary) *redactionSummaryCanonicalV1 {
	if in == nil {
		return nil
	}
	return &redactionSummaryCanonicalV1{
		Profile:           in.Profile,
		Provider:          in.Provider,
		Parser:            in.Parser,
		TotalRedactions:   in.TotalRedactions,
		ByClass:           in.ByClass,
		CacheBoundaryKept: in.CacheBoundaryKept,
	}
}

type shieldSummaryCanonicalV1 struct {
	Pipeline                 string `json:"pipeline,omitempty"`
	TotalRewrites            int    `json:"total_rewrites,omitempty"`
	ExtensionProbes          int    `json:"extension_probes,omitempty"`
	TrackingBeacons          int    `json:"tracking_beacons,omitempty"`
	AgentTraps               int    `json:"agent_traps,omitempty"`
	FingerprintShimInjected  bool   `json:"fingerprint_shim_injected,omitempty"`
	SVGForeignObjects        int    `json:"svg_foreign_objects,omitempty"`
	SVGEventHandlers         int    `json:"svg_event_handlers,omitempty"`
	SVGExternalReferences    int    `json:"svg_external_references,omitempty"`
	SVGHiddenText            int    `json:"svg_hidden_text,omitempty"`
	SVGAnimationInjections   int    `json:"svg_animation_injections,omitempty"`
	BodyBytes                int    `json:"body_bytes,omitempty"`
	ScannedBytes             int    `json:"scanned_bytes,omitempty"`
	Partial                  bool   `json:"partial,omitempty"`
	AdaptiveSignalsRecorded  int    `json:"adaptive_signals_recorded,omitempty"`
	AdaptiveSignalMaxPerBody int    `json:"adaptive_signal_max_per_body,omitempty"`
}

func canonicalShieldSummaryV1(in *ShieldSummary) *shieldSummaryCanonicalV1 {
	if in == nil {
		return nil
	}
	return &shieldSummaryCanonicalV1{
		Pipeline:                 in.Pipeline,
		TotalRewrites:            in.TotalRewrites,
		ExtensionProbes:          in.ExtensionProbes,
		TrackingBeacons:          in.TrackingBeacons,
		AgentTraps:               in.AgentTraps,
		FingerprintShimInjected:  in.FingerprintShimInjected,
		SVGForeignObjects:        in.SVGForeignObjects,
		SVGEventHandlers:         in.SVGEventHandlers,
		SVGExternalReferences:    in.SVGExternalReferences,
		SVGHiddenText:            in.SVGHiddenText,
		SVGAnimationInjections:   in.SVGAnimationInjections,
		BodyBytes:                in.BodyBytes,
		ScannedBytes:             in.ScannedBytes,
		Partial:                  in.Partial,
		AdaptiveSignalsRecorded:  in.AdaptiveSignalsRecorded,
		AdaptiveSignalMaxPerBody: in.AdaptiveSignalMaxPerBody,
	}
}
